sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XXE/XmlReaderSettingsValidator.cs

```diff
@@ -0,0 +1,106 @@
+        }
+
+        public bool IsUnsafe(InvocationExpressionSyntax invocation, ISymbol settings, NetFrameworkVersion version)
+        {
+            var unsafeDtdProcessing = false;
+            var unsafeResolver = version == NetFrameworkVersion.Probably35 || version == NetFrameworkVersion.Between4And451;
+
+            var objectCreation = GetObjectCreation(settings, invocation, semanticModel);
+            var objectCreationAssignments = objectCreation.GetInitializerExpressions().OfType<AssignmentExpressionSyntax>();
+
+            var assignments = invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>()
+                .DescendantNodes()
+                .OfType<AssignmentExpressionSyntax>()
+                .Where(assignment => assignment.Left is MemberAccessExpressionSyntax memberAccess &&
+                                     IsXmlReaderSettings(memberAccess.Expression, semanticModel) &&
+                                     settings.Equals(semanticModel.GetSymbolInfo(memberAccess.Expression).Symbol));
+
+            foreach (var assignment in objectCreationAssignments.Union(assignments))
+            {
+                var name = assignment.Left.GetName();
+
+                if (name =="ProhibitDtd" || name == "DtdProcessing")
+                {
+                    unsafeDtdProcessing = IsXmlResolverDtdProcessingUnsafe(assignment, semanticModel);
+                }
+                else if (assignment.Left.NameIs("XmlResolver"))
+                {
+                    unsafeResolver = IsXmlResolverAssignmentUnsafe(assignment, semanticModel);
+                }
+            }
+
+            return unsafeDtdProcessing && unsafeResolver;
+        }
+
+        private static bool IsXmlReaderSettings(ExpressionSyntax expressionSyntax, SemanticModel semanticModel) =>
+            semanticModel.GetTypeInfo(expressionSyntax).Type.Is(KnownType.System_Xml_XmlReaderSettings);
+
+        private static ObjectCreationExpressionSyntax GetObjectCreation(ISymbol symbol, InvocationExpressionSyntax invocation, SemanticModel semanticModel) =>
+            symbol.Locations
+                .SelectMany(location => GetDescendantNodes(location, invocation).OfType<ObjectCreationExpressionSyntax>())
+                .FirstOrDefault(objectCreation => objectCreation.Initializer != null && IsXmlReaderSettings(objectCreation, semanticModel));
+
+        private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, InvocationExpressionSyntax invocation) =>
+            location.SourceTree?.GetRoot()?.FindNode(location.SourceSpan).DescendantNodes() ??
+            invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>()?.DescendantNodes() ??
+            Enumerable.Empty<SyntaxNode>();
+
+        private static bool IsXmlResolverDtdProcessingUnsafe(AssignmentExpressionSyntax assignment, SemanticModel semanticModel) =>
+            semanticModel.GetConstantValue(assignment.Right).Value switch
+            {
+                false => true,
```

Comment:
small comment above that this `false` is for `ProhibitDtd`?

`false =>  true` is inherently difficult to grasp :)

Reply:
😄 indeed